### ISSUE TITLE
unnecessary check aws_vpc_ipam_pool_invalid_aws_service causing unnecessary errors

### ISSUE THREAD

**Issue by drewmullen** [terraform-linters/tflint-ruleset-aws#357]
My module is getting an error saying:

> Failed to check ruleset; Failed to check `aws_vpc_ipam_pool_invalid_aws_service` rule: failed to eval an expression in modules/sub_pool/main.tf:24; Unsupported attribute: This object does not have an attribute named "aws_service".

Hoping someone can clarify what is going on with the error... [we do provide the argument](https://github.com/aws-ia/terraform-aws-ipam/blob/a7d508cb0be2f68d99952682c2392b6d7d541d96/modules/sub_pool/main.tf#L24) value as either user provider or `null`

That being said, this check is unnecessary since the service values are[ validated during `terraform plan`](https://github.com/hashicorp/terraform-provider-aws/blob/b9fd7aba413d3967d89f8d873432f910e5905bea/internal/service/ec2/ipam_pool.go#L68) (no API interaction required)

**Comment by drewmullen** [comment]
its right here: https://github.com/aws-ia/terraform-aws-ipam/blob/4617b236e4ef2accf2616595e55652139dcbdc59/modules/sub_pool/variables.tf#L12

**Comment by wata727** [comment]
If the value is null, you got the "Attempt to get attribute from null value" error. You're probably passing a value of a different object type.

TFLint is different from validation by the provider. I believe there are many situations where TFLint is useful, but if you feel you don't need it, you can disable that rule.

**Comment by drewmullen** [comment]
> I believe there are many situations where TFLint is useful 

agreed! i use tflint for a lot of things such as ensuring version pinning and we block out several resource types that we dont want people to use

If i can just push a little bit more to try and understand... what is the use case for this check? the test only validates that youre passing an approved string value and it uses a static set. 

To me this check in tflint is the equivalent of writing a variable validation for something like that... eg:

```terraform
variable "aws_service" {
  validation {
    error_message = "Must only pass valid values. Valid values: \"ec2\"."
    condition = var.aws_service == "ec2" || var.aws_service == null
  }
}
```

We would never do that because 

1. its already covered by the provider 
1. its manually expressed in my codebase now

---

if youre going to keep the check in, i recommend you reference the allowed values function from the SDK instead of hardcoding the approved values in. These are planned to be expanded overtime. 

Heres the function for the aws-sdk-go
```
// IpamPoolAwsService_Values returns all elements of the IpamPoolAwsService enum
func IpamPoolAwsService_Values() []string {
	return []string{
		IpamPoolAwsServiceEc2,
	}
}
```

--- 

Back to the error:

I think i figured out what is giving tflint heartburn and it may be a broader problem. [I do not pass the value in 1 of the module references](https://github.com/aws-ia/terraform-aws-ipam/blob/4617b236e4ef2accf2616595e55652139dcbdc59/main.tf#L49-L56). Terraform then views that value as `null` because because var.pool_config in the module has `aws_service` as `optional()` aka `null`

If i update that module reference to include `pool_config = { ... aws_service = null}` explicitly then tflint no longer errors. 

I believe tflint is missing the implicit null from the optional() setting - thoughts?

**Comment by wata727** [comment]
Some rules in TFLint are automatically generated from the AWS OpenAPI specification, and this rule is also a part of this. This is the same mechanism as the automatic generation of aws-sdk-go, and if there is an update, it will be updated automatically.

> I believe tflint is missing the implicit null from the optional() setting - thoughts?

Umm, you're right. I have confirmed that using optional object attributes gives different results than Terraform.

```hcl
terraform {
  experiments = [module_variable_optional_attrs]
}

variable "foo" {
  type = object({
    a = optional(string)
    b = optional(string)
  })
}

resource "aws_instance" "foo" {
  instance_type = var.foo.b
}
```

```console
% TF_VAR_foo={a=1} tflint
Failed to check ruleset; Failed to check `aws_instance_previous_type` rule: failed to eval an expression in main.tf:13; Unsupported attribute: This object does not have an attribute named "b".
```

```console
% TF_VAR_foo={a=1} terraform console
╷
│ Warning: Experimental feature "module_variable_optional_attrs" is active
│
│   on main.tf line 2, in terraform:
│    2:   experiments = [module_variable_optional_attrs]
│
│ Experimental features are subject to breaking changes in future minor or patch releases, based on feedback.
│
│ If you have feedback on the design of this feature, please open a GitHub issue to discuss it.
╵

> var.foo.b
tostring(null)
```

**Comment by wata727** [comment]
https://github.com/terraform-linters/tflint/issues/1436

**Comment by drewmullen** [comment]
> it will be updated automatically.

Good to know. The check itself is not really a problem now that I know it's a broader bug. We can close this out. Thanks for helping investigated

**Comment by drewmullen** [comment]
tflint v39.1 + 0.15 aws-ruleset solved this issue for me. thank you!
